### ISSUE TITLE
구현 과정

### ISSUE THREAD

**Issue by Jeeehee** [Jeeehee/HandLink#6]
## 1. ML Framework
Google의 `MediaPipe` Hand는 모든 이미지에서 작동하고 3D Hand Keypoint로 높은 충실도를 갖는 21개의 3D LandMark를 추출합니다.
부분적으로 보이는 손이나, 겹치는 손도 인식해주어 보다 정확하게 수어를 번역할 수 있다고 판단하였고, ML Framework로 선택하게 되었습니다.

<br>

## 2. 구현 언어
ML Framework를 MediaPipe로 결정 후,  Swift를 이용해 딥러닝을 구현하고자 하였습니다.  
하지만, MediaPipe github의 [해당 문서](https://github.com/google/mediapipe/blob/master/mediapipe/objc/MPPGraph.h)의 19번째 줄을 보면, 아래와 같이 명시되어있습니다.  
  ```
  #error This header can only be included by an Objective-C++ file.
  ```
따라서 Swift가 아닌 다른 언어로 진행해야 했고, Objective-C++ 보다 쉽고 자료가 많은 **Python을 선택**하게 되었습니다.

<br>

## 3. DataSet 생성
  1. `OpenCV` 로 이미지를 실시간으로 프로세싱
  2. `MediaPipe`로 [x, y, z]와, LandMark가 이미지 상에 보이는지 확인하는 [visibility]까지 총 4개의 좌표로 관절 사이의 각도를 계산
  3. 3D LandMark를 추출
  4. index로 학습 언어 label 부여
  5. npy 형태로 DataSet 저장
  
<br>

## 4. 모델 생성 및 학습
  1. 앞서 추출한 DataSet을 하나로 병합
  2. `tensorflow의 keras` 로 One-Hot-Encoding
  3. `sklearn` 을 사용해 트레이닝 데이터셋과, 테스트 데이터 셋을 9:1로 나눔
  4. `Sequential` 을 사용해 모델을 정의, 44,229개의 파라미터로 학습 진행  

       <img width="350" alt="스크린샷 2022-11-08 오전 10 31 06" src="https://user-images.githubusercontent.com/92635121/200451815-e1509bf2-04ce-466d-9b9c-8e94eadcf5ff.png">
  5. 세부 사항 및 결과는 [해당 링크](https://github.com/Jeeehee/HandLink/issues/2)에서 확인 가능합니다.
  
    
<br>

## 5. Test
  1. 90% 이상의 정확도를 가진 제스처만 인식하여, Action Sequece에 저장
  2. 마지막 3개의 Action이 동일할 때만 유효한 액션이라고 판단
  3. 판단 완료 된 프레임에 해당 수어의 뜻 출력
  
      <img width="350" alt="스크린샷 2022-11-07 오후 5 13 14" src="https://user-images.githubusercontent.com/92635121/202601483-82109651-a34b-44d0-afb6-e46e93ce9a5a.png">
